contributors/design-proposals/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,570 @@
+
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
+reattach to Debug Container if you've been disconnected. This is accomplished by
+running `kubectl debug -r` with the name of an existing, running Debug
+Container. The `-r` option to `kubectl debug` is translated to a `reattach=True`
+parameter. It is an error to attempt to reattach to a Debug Container that does
+not exist or is not running.
+
+Action                   | Exiting Container State | Result
+------------------------ | ----------------------- | ----------------
+Create Debug Container   | Does Not Exist          | Create & Attach
+"                        | Running                 | *ERROR*
+"                        | Exited                  | Replace & Attach
+Reattach Debug Container | Does Not Exist          | *ERROR*
+"                        | Running                 | Attach
+"                        | Exited                  | *ERROR*
+
+When supported by a runtime, multiple clients can attach to a single debug
+container and share the terminal. This is supported by Docker.
+
+### Killing Debug Containers
+
+Debug containers will not be killed automatically until the pod (specifically,
+the pod sandbox) is destroyed. Unlike `kubectl exec`, Debug Containers will not
+receive an EOF if their connection is interrupted. Instead, Debug Containers
+must be reattached to exit a running process. This could be tricky if the
+process does not allocate a TTY, in this case a second Debug Container could be
+used to deliver a signal via `kill(1)`.
+
+### Container Lifecycle Changes
+
+Implementing debug requires no changes to the Container Runtime Interface as
+it's the same operation as creating a regular container. The following changes
+are necessary in the kubelet:
+
+1.  `SyncPod()` must not kill any Debug Container even though it is not part of
+    the pod spec.
+1.  As an exception to the above, `SyncPod()` will kill Debug Containers when
+    the pod sandbox changes since a Debug Container in a no-longer-used sandbox
+    is not useful. Debug Containers are not automatically restarted in the new
+    sandbox.
+1.  `convertStatusToAPIStatus()` must sort Debug Containers status into
+    `DebugContainerStatuses` similar to as it does for `InitContainerStatuses`
+1.  The kubelet must preserve information on debug containers for reporting.
+1.  Debug Containers must be excluded from calculation of pod phase and
+    condition
+
+It's worth noting some things that do not need to change:
+
+1.  `KillPod()` already operates on all running containers returned by the
+    runtime.
+1.  Containers created prior to this feature being enabled will have a
+    `containerType` of `""`. Since this does not match `"DEBUG"` the special
+    handling of Debug Containers is backwards compatible.
+
+### Additional Constraints
+
+1.  Non-interactive workloads are explicitly supported. There are no plans to
```

Comment:
If the container run and attach steps are distinct, how is stdout/stderr coordinated so that the attach request obtains the first byte written to each? Is the first attach special? How do subsequent or additional attaches behave w.r.t. previous output from the debug container's entrypoint?